odoo/tools/misc.py

```diff
@@ -1254,13 +1254,55 @@ def __gt__(self, other): return self.val < other.val
     def __le__(self, other): return self.val >= other.val
     def __lt__(self, other): return self.val > other.val
 
+# there is contextlib.suppress since py3
 @contextmanager
 def ignore(*exc):
     try:
         yield
     except exc:
         pass
 
+class reraise_x_as(ContextDecorator):
+    """
+    Hide some exceptions behind a fake alternative error. Can be used as
+    a function decorator or as a context manager.
+
+    .. code-block:
+
+        @route('/super/secret/route', auth='public')
+        @reraise_x_as(AccessError, as_=NotFound())
+        def super_secret_route(self):
+            if not request.session.uid:
+                raise AccessError("Route hidden to non logged-in users")
+            ...
+
+        def some_util():
+            ...
+            with reraise_x_as(ValueError, as_=UserError("Invalid argument")):
+                ...
+            ...
+
+    :param exceptions: the exception classes to except.
+    :param as_: the fake exception to raise instead.
+    """
+    def __init__(self, *exceptions, as_):
+        if not exceptions:
+            raise ValueError("Missing exceptions")
+
+        wrong_exc = next((not issubclass(exc, Exception) for exc in exceptions), None)
+        if wrong_exc:
+            raise TypeError(f"{wrong_exc} is not an exception class.")
+
+        self.exceptions = exceptions
+        self.as_ = as_
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if exc_type is not None and issubclass(exc_type, self.exceptions):
+            raise self.as_ from exc_value
```

Comment:
This seems pretty risky as in decorator mode this will be reusing the same exception object over and over again. I don't know how well supported that is.

Maybe the substitution should be a callable?

Reply:
To be fair I don't think it is a problem to use the same exception multiple times

Reply:
> To be fair I don't think it is a problem to use the same exception multiple times

I don't know about that, exceptions are mutable objects which hold references to their raising environment (traceback, cause, ...), and sometimes we carry them around for a while (e.g. `_handle_exception`). So in threaded mode, depending how much this decorator / context manager gets used we might get very strange behaviours if multiple requests conflict on it.